proposals/utils/stepper.py

```diff
@@ -0,0 +1,99 @@
+from proposals.models import Proposal
+from django.urls import reverse
+from django.template import loader, Template
+from django.core.exceptions import ImproperlyConfigured
+
+def flatten(lst):
+    if lst == []:
+        return lst
+    first_item = lst[0]
+    rest = lst[1:]
+    if type(first_item) is list:
+        return flatten(first_item) + flatten(rest)
+    return [first_item] + flatten(rest)
+
+
+class renderable:
+
+    def get_context_data(self):
+        return {}
+
+    def render(self, context={}):
+        template = loader.get_template(self.template_name)
+        context.update(self.get_context_data())
+        return template.render(context.flatten())
+
+
+class Stepper(renderable):
+
+    template_name = "base/stepper.html"
+
+    def __init__(self, proposal, current=None):
+
+        self.proposal = proposal
+        self.current = current
+
+    def get_context_data(self):
+        return {"stepper": self}
+
+    def get_resume_url(self):
+        """
+        Returns the url of the first page that requires attention,
+        that being either a page with an error or an incomplete page.
+        """
+        pages = self.collect_pages()
+
+        for page in flatten(pages):
+            if not page.is_complete:
+                return page.get_url()
+
+    def collect_items(self):
+        some_pages = []
+        for i in range(5):
+            some_pages.append(
+                Page(self.proposal)
+            )
+        return some_pages
+
+
+class StepperContextMixin:
+    """
+    Includes a stepper object in the view's context
+    """
+
+    def get_context_data(self, *args, **kwargs):
+        context = super().get_context_data(*args, **kwargs)
+        # Try to determine proposal
+        if hasattr(self, "proposal"):
+            proposal = self.proposal
+        else:
+            proposal = Proposal.objects.get(pk=self.kwargs.get("pk"))
+        # Initialize and insert stepper object
+        context["stepper"] = Stepper(proposal)
+        return context
```

Comment:
I think you know this as well, but this won't work for any page that does not operate on Proposal objects.

Maybe we could extract some code from [`UserAllowedMixin` ](https://github.com/DH-IT-Portal-Development/ethics/blob/8ac03c15561b1eb8068499d0eeeee0b4845ecb23/main/views.py#L339) and add it to a `ProposalMixin` (which, annoyingly, already exists, but that should be renamed)

Reply:
Also, as an aside, it _might_ be an idea to create the stepper object in `Proposal`, as we may also want to use the data from it as part of the logic of the app. Then, we add proposal to the context instead and call `proposal.stepper` in the template.

This is just me brainstorming, the current setup is fine with me too

Reply:
> I think you know this as well, but this won't work for any page that does not operate on Proposal objects. 

Yeah I know, it's early days. A proposal object is all a stepper should ever need to generate itself. Determining the proposal from any of the sub-objects is quite trivial, but I'm not sure whose responsibility it is.

I could grant the stepper the power to initialize from any of the sub-objects ("DWIM"). ALternatively I could make a method in StepperContextMixin called `determine_proposal_from_object()` that can figure it out. The latter could prove useful to have lying around.

We could also have every proposal view be separately responsible for setting `self.proposal` as a rule. But that feels brittle.